File: photo_sorter.py
import os
import shutil
import hashlib
from pathlib import Path
from datetime import datetime

class PhotoSorter:
    def __init__(self, recognizer, db, base_output_dir="sorted_photos"):
        self.recognizer = recognizer
        self.db = db
        self.base_output_dir = base_output_dir
        
        Path(base_output_dir).mkdir(exist_ok=True)
        
        # Создаем папки для разных категорий
        self.unknown_dir = os.path.join(base_output_dir, "unknown")
        self.no_faces_dir = os.path.join(base_output_dir, "no_faces")
        Path(self.unknown_dir).mkdir(exist_ok=True)
        Path(self.no_faces_dir).mkdir(exist_ok=True)
    
    def get_file_hash(self, file_path):
        """Вычисление MD5 хеша файла"""
        hasher = hashlib.md5()
        with open(file_path, 'rb') as f:
            for chunk in iter(lambda: f.read(65536), b''):
                hasher.update(chunk)
        return hasher.hexdigest()
    
    def is_duplicate(self, file_path, target_dir):
        """Проверка, есть ли уже такой файл в целевой папке"""
        file_hash = self.get_file_hash(file_path)
        
        if not os.path.exists(target_dir):
            return False, None
        
        for existing_file in os.listdir(target_dir):
            existing_path = os.path.join(target_dir, existing_file)
            if os.path.isfile(existing_path):
                try:
                    existing_hash = self.get_file_hash(existing_path)
                    if existing_hash == file_hash:
                        return True, existing_path
                except:
                    continue
        return False, None
    
    def sort_photo(self, image_path):
        """
        Умная сортировка одного фото
        """
        print(f"\n📁 Обработка: {os.path.basename(image_path)}")
        
        if not os.path.exists(image_path):
            return {"status": "error", "message": "Файл не найден"}
        
        # Анализируем фото
        results = self.recognizer.process_image(image_path)
        
        if not results:
            # Нет лиц
            is_dup, existing = self.is_duplicate(image_path, self.no_faces_dir)
            if is_dup:
                print(f"   ⏭️ Дубликат (уже есть в no_faces)")
                return {"status": "duplicate", "dest_path": existing}
            
            dest_path = self._copy_file(image_path, self.no_faces_dir)
            return {"status": "no_faces", "dest_path": dest_path}
        
        # Сортируем по лицам
        sorted_info = []
        for face_data in results:
            person_name = face_data["name"]
            confidence = face_data["confidence"]
            
            if person_name == "Неизвестный":
                person_dir = self.unknown_dir
            else:
                person_dir = os.path.join(self.base_output_dir, person_name)
                Path(person_dir).mkdir(exist_ok=True, parents=True)
            
            # Проверяем дубликат
            is_dup, existing = self.is_duplicate(image_path, person_dir)
            if is_dup:
                print(f"   ⏭️ Дубликат {person_name} (уже есть)")
                sorted_info.append({
                    "person": person_name,
                    "confidence": confidence,
                    "dest_path": existing,
                    "action": "duplicate"
                })
                continue
            
            # Копируем файл
            dest_path = self._copy_file(image_path, person_dir)
            print(f"   ✅ {person_name} -> {os.path.basename(dest_path)}")
            
            # Сохраняем в БД
            self.db.save_sorted_photo(image_path, dest_path, person_name, confidence)
            
            sorted_info.append({
                "person": person_name,
                "confidence": confidence,
                "dest_path": dest_path,
                "action": "copied"
            })
        
        return {"status": "sorted", "sorted_info": sorted_info}
    
    def _copy_file(self, src_path, dest_dir):
        """Копирует файл с уникальным именем на основе хеша"""
        file_hash = self.get_file_hash(src_path)
        ext = os.path.splitext(src_path)[1]
        
        # Используем хеш как имя файла для избежания дубликатов
        dest_path = os.path.join(dest_dir, f"{file_hash}{ext}")
        
        if not os.path.exists(dest_path):
            shutil.copy2(src_path, dest_path)
        
        return dest_path
    
    def resort_all_photos(self):
        """Пересортировать ВСЕ фото из unknown на основе новых знаний"""
        print("\n" + "="*60)
        print("🔄 ПЕРЕСОРТИРОВКА ВСЕХ ФОТО (новые знания)")
        print("="*60)
        
        if not os.path.exists(self.unknown_dir):
            print("📁 Папка unknown не найдена")
            return []
        
        # Находим все фото в unknown
        unknown_photos = []
        for file in os.listdir(self.unknown_dir):
            if file.lower().endswith(('.jpg', '.jpeg', '.png', '.bmp')):
                unknown_photos.append(os.path.join(self.unknown_dir, file))
        
        print(f"📸 Найдено {len(unknown_photos)} фото в папке unknown")
        
        results = []
        for photo_path in unknown_photos:
            print(f"\n🔍 Проверяем: {os.path.basename(photo_path)}")
            
            # Анализируем заново
            analysis = self.recognizer.process_image(photo_path)
            
            if analysis:
                for face_data in analysis:
                    person_name = face_data["name"]
                    confidence = face_data["confidence"]
                    
                    if person_name != "Неизвестный" and confidence > 0.5:
                        # О! Теперь мы знаем кто это!
                        print(f"   ✅ РАСПОЗНАН: {person_name} (уверенность: {confidence:.2%})")
                        
                        # Создаем папку для человека
                        person_dir = os.path.join(self.base_output_dir, person_name)
                        Path(person_dir).mkdir(exist_ok=True, parents=True)
                        
                        # Проверяем дубликат
                        is_dup, existing = self.is_duplicate(photo_path, person_dir)
                        if is_dup:
                            print(f"   ⏭️ Фото уже есть в папке {person_name}")
                            # Удаляем из unknown
                            os.remove(photo_path)
                            results.append({
                                "file": os.path.basename(photo_path),
                                "action": "removed_duplicate",
                                "person": person_name
                            })
                        else:
                            # Копируем в папку человека
                            dest_path = self._copy_file(photo_path, person_dir)
                            print(f"   📁 ПЕРЕМЕЩЕНО: {dest_path}")
                            
                            # Сохраняем в БД
                            self.db.save_sorted_photo(photo_path, dest_path, person_name, confidence)
                            
                            # Удаляем из unknown
                            os.remove(photo_path)
                            results.append({
                                "file": os.path.basename(photo_path),
                                "action": "moved_to_known",
                                "person": person_name,
                                "dest_path": dest_path
                            })
                        break
                    else:
                        print(f"   ⚠️ Остается неизвестным (уверенность: {confidence:.2%})")
            else:
                # Нет лиц - перемещаем в no_faces
                no_faces_path = os.path.join(self.no_faces_dir, os.path.basename(photo_path))
                if not os.path.exists(no_faces_path):
                    shutil.move(photo_path, no_faces_path)
                    print(f"   📁 Перемещено в no_faces")
                    results.append({
                        "file": os.path.basename(photo_path),
                        "action": "moved_to_no_faces"
                    })
        
        print("\n" + "="*60)
        print(f"✅ ПЕРЕСОРТИРОВКА ЗАВЕРШЕНА!")
        print(f"📊 Распознано новых лиц: {len([r for r in results if r['action'] == 'moved_to_known'])}")
        print("="*60 + "\n")
        
        return results
    
    def sort_folder(self, folder_path, progress_callback=None):
        """Сортировка новой папки (без дубликатов)"""
        image_extensions = {'.jpg', '.jpeg', '.png', '.bmp', '.tiff'}
        results = []
        
        # Находим все фото
        photos = []
        for file in os.listdir(folder_path):
            if os.path.splitext(file)[1].lower() in image_extensions:
                photos.append(os.path.join(folder_path, file))
        
        total = len(photos)
        print(f"\n📸 Найдено {total} новых фото")
        
        for i, photo_path in enumerate(photos):
            if progress_callback:
                progress_callback(i + 1, total, os.path.basename(photo_path))
            
            result = self.sort_photo(photo_path)
            results.append({
                "file": os.path.basename(photo_path),
                "result": result
            })
        
        return results

File: face_organizer/lib/services/photo_sorter.dart
import 'dart:io';
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import 'package:path/path.dart' as path;
import 'package:crypto/crypto.dart';
import '../database/face_database.dart';
import 'face_recognizer.dart';

class PhotoSorter {
  final FaceRecognizer recognizer;
  final FaceDatabase db;
  final String baseOutputDir;
  late String unknownDir;
  late String noFacesDir;
  
  PhotoSorter(this.recognizer, this.db, this.baseOutputDir) {
    _createDirectories();
  }
  
  void _createDirectories() {
    Directory(baseOutputDir).createSync(recursive: true);
    
    unknownDir = path.join(baseOutputDir, "unknown");
    noFacesDir = path.join(baseOutputDir, "no_faces");
    
    Directory(unknownDir).createSync(recursive: true);
    Directory(noFacesDir).createSync(recursive: true);
  }
  
  Future<String> getFileHash(String filePath) async {
    try {
      final file = File(filePath);
      final bytes = await file.readAsBytes();
      final hash = md5.convert(bytes);
      return hash.toString();
    } catch (e) {
      print("Ошибка вычисления хеша: $e");
      return "";
    }
  }
  
  Future<(bool, String?)> isDuplicate(String filePath, String targetDir) async {
    final targetDirectory = Directory(targetDir);
    if (!await targetDirectory.exists()) {
      return (false, null);
    }
    
    final fileHash = await getFileHash(filePath);
    if (fileHash.isEmpty) return (false, null);
    
    final files = await targetDirectory.list().where((entity) => entity is File).toList();
    
    for (var entity in files) {
      final existingFile = entity as File;
      try {
        final existingHash = await getFileHash(existingFile.path);
        if (existingHash == fileHash) {
          return (true, existingFile.path);
        }
      } catch (e) {
        continue;
      }
    }
    
    return (false, null);
  }
  
  Future<String> copyFile(String srcPath, String destDir) async {
    final fileHash = await getFileHash(srcPath);
    final ext = path.extension(srcPath);
    
    final destPath = path.join(destDir, "$fileHash$ext");
    final destFile = File(destPath);
    
    if (!await destFile.exists()) {
      await File(srcPath).copy(destPath);
    }
    
    return destPath;
  }
  
  Future<Map<String, dynamic>> sortPhoto(String imagePath) async {
    print("\n📁 Обработка: ${path.basename(imagePath)}");
    
    final imageFile = File(imagePath);
    if (!await imageFile.exists()) {
      return {"status": "error", "message": "Файл не найден"};
    }
    
    // Анализируем фото
    final results = await recognizer.processImage(imagePath, (faceData) {});
    
    if (results.isEmpty) {
      // Нет лиц
      final (isDup, existing) = await isDuplicate(imagePath, noFacesDir);
      if (isDup) {
        print("   ⏭️ Дубликат (уже есть в no_faces)");
        return {"status": "duplicate", "dest_path": existing};
      }
      
      final destPath = await copyFile(imagePath, noFacesDir);
      return {"status": "no_faces", "dest_path": destPath};
    }
    
    // Сортируем по лицам
    final sortedInfo = <Map<String, dynamic>>[];
    
    for (var faceData in results) {
      final personName = faceData["name"] as String;
      final confidence = faceData["confidence"] as double;
      
      String personDir;
      if (personName == "Неизвестный") {
        personDir = unknownDir;
      } else {
        personDir = path.join(baseOutputDir, personName);
        await Directory(personDir).create(recursive: true);
      }
      
      // Проверяем дубликат
      final (isDup, existing) = await isDuplicate(imagePath, personDir);
      if (isDup) {
        print("   ⏭️ Дубликат $personName (уже есть)");
        sortedInfo.add({
          "person": personName,
          "confidence": confidence,
          "dest_path": existing,
          "action": "duplicate"
        });
        continue;
      }
      
      // Копируем файл
      final destPath = await copyFile(imagePath, personDir);
      print("   ✅ $personName -> ${path.basename(destPath)}");
      
      // Сохраняем в БД
      await db.saveSortedPhoto(imagePath, destPath, personName, confidence);
      
      sortedInfo.add({
        "person": personName,
        "confidence": confidence,
        "dest_path": destPath,
        "action": "copied"
      });
    }
    
    return {"status": "sorted", "sorted_info": sortedInfo};
  }
  
  Future<List<Map<String, dynamic>>> resortAllPhotos() async {
    print("\n" + "=" * 60);
    print("🔄 ПЕРЕСОРТИРОВКА ВСЕХ ФОТО (новые знания)");
    print("=" * 60);
    
    final unknownDirectory = Directory(unknownDir);
    if (!await unknownDirectory.exists()) {
      print("📁 Папка unknown не найдена");
      return [];
    }
    
    // Находим все фото в unknown
    final unknownPhotos = <String>[];
    final files = await unknownDirectory.list().where((entity) => entity is File).toList();
    
    for (var file in files) {
      final fileName = path.basename(file.path).toLowerCase();
      if (fileName.endsWith('.jpg') || fileName.endsWith('.jpeg') || 
          fileName.endsWith('.png') || fileName.endsWith('.bmp')) {
        unknownPhotos.add(file.path);
      }
    }
    
    print("📸 Найдено ${unknownPhotos.length} фото в папке unknown");
    
    final results = <Map<String, dynamic>>[];
    
    for (var photoPath in unknownPhotos) {
      print("\n🔍 Проверяем: ${path.basename(photoPath)}");
      
      // Анализируем заново
      final analysis = await recognizer.processImage(photoPath, (faceData) {});
      
      if (analysis.isNotEmpty) {
        bool recognized = false;
        for (var faceData in analysis) {
          final personName = faceData["name"] as String;
          final confidence = faceData["confidence"] as double;
          
          if (personName != "Неизвестный" && confidence > 0.5) {
            print("   ✅ РАСПОЗНАН: $personName (уверенность: ${(confidence * 100).toStringAsFixed(1)}%)");
            
            // Создаем папку для человека
            final personDir = path.join(baseOutputDir, personName);
            await Directory(personDir).create(recursive: true);
            
            // Проверяем дубликат
            final (isDup, existing) = await isDuplicate(photoPath, personDir);
            if (isDup) {
              print("   ⏭️ Фото уже есть в папке $personName");
              // Удаляем из unknown
              await File(photoPath).delete();
              results.add({
                "file": path.basename(photoPath),
                "action": "removed_duplicate",
                "person": personName
              });
            } else {
              // Копируем в папку человека
              final destPath = await copyFile(photoPath, personDir);
              print("   📁 ПЕРЕМЕЩЕНО: $destPath");
              
              // Сохраняем в БД
              await db.saveSortedPhoto(photoPath, destPath, personName, confidence);
              
              // Удаляем из unknown
              await File(photoPath).delete();
              results.add({
                "file": path.basename(photoPath),
                "action": "moved_to_known",
                "person": personName,
                "dest_path": destPath
              });
            }
            recognized = true;
            break;
          } else {
            print("   ⚠️ Остается неизвестным (уверенность: ${(confidence * 100).toStringAsFixed(1)}%)");
          }
        }
        if (!recognized) {
          // Остается неизвестным - ничего не делаем
        }
      } else {
        // Нет лиц - перемещаем в no_faces
        final noFacesPath = path.join(noFacesDir, path.basename(photoPath));
        final noFacesFile = File(noFacesPath);
        if (!await noFacesFile.exists()) {
          await File(photoPath).rename(noFacesPath);
          print("   📁 Перемещено в no_faces");
          results.add({
            "file": path.basename(photoPath),
            "action": "moved_to_no_faces"
          });
        }
      }
    }
    
    print("\n" + "=" * 60);
    print("✅ ПЕРЕСОРТИРОВКА ЗАВЕРШЕНА!");
    final movedCount = results.where((r) => r['action'] == 'moved_to_known').length;
    print("📊 Распознано новых лиц: $movedCount");
    print("=" * 60 + "\n");
    
    return results;
  }
  
  Future<List<Map<String, dynamic>>> sortFolder(String folderPath, Function(int, int, String)? progressCallback) async {
    final imageExtensions = {'.jpg', '.jpeg', '.png', '.bmp', '.tiff'};
    final photos = <String>[];
    
    final folder = Directory(folderPath);
    final files = await folder.list().where((entity) => entity is File).toList();
    
    for (var file in files) {
      final ext = path.extension(file.path).toLowerCase();
      if (imageExtensions.contains(ext)) {
        photos.add(file.path);
      }
    }
    
    final total = photos.length;
    print("\n📸 Найдено $total новых фото");
    
    final results = <Map<String, dynamic>>[];
    
    for (int i = 0; i < total; i++) {
      final photoPath = photos[i];
      if (progressCallback != null) {
        progressCallback(i + 1, total, path.basename(photoPath));
      }
      
      final result = await sortPhoto(photoPath);
      results.add({
        "file": path.basename(photoPath),
        "result": result
      });
    }
    
    return results;
  }
}
